File: app/src/main/java/org/schabi/newpipe/player/playqueue/PlayQueueItem.java
package org.schabi.newpipe.player.playqueue;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.schabi.newpipe.extractor.Image;
import org.schabi.newpipe.extractor.stream.StreamInfo;
import org.schabi.newpipe.extractor.stream.StreamInfoItem;
import org.schabi.newpipe.extractor.stream.StreamType;
import org.schabi.newpipe.util.ExtractorHelper;

import java.io.Serializable;
import java.util.List;

import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class PlayQueueItem implements Serializable {
    public static final long RECOVERY_UNSET = Long.MIN_VALUE;
    private static final String EMPTY_STRING = "";

    @NonNull
    private final String title;
    @NonNull
    private final String url;
    private final int serviceId;
    private final long duration;
    @NonNull
    private final List<Image> thumbnails;
    @NonNull
    private final String uploader;
    private final String uploaderUrl;
    @NonNull
    private final StreamType streamType;

    private boolean isAutoQueued;

    private long recoveryPosition;
    private Throwable error;

    PlayQueueItem(@NonNull final StreamInfo info) {
        this(info.getName(), info.getUrl(), info.getServiceId(), info.getDuration(),
                info.getThumbnails(), info.getUploaderName(),
                info.getUploaderUrl(), info.getStreamType());

        if (info.getStartPosition() > 0) {
            setRecoveryPosition(info.getStartPosition() * 1000);
        }
    }

    PlayQueueItem(@NonNull final StreamInfoItem item) {
        this(item.getName(), item.getUrl(), item.getServiceId(), item.getDuration(),
                item.getThumbnails(), item.getUploaderName(),
                item.getUploaderUrl(), item.getStreamType());
    }

    @SuppressWarnings("ParameterNumber")
    private PlayQueueItem(@Nullable final String name, @Nullable final String url,
                          final int serviceId, final long duration,
                          final List<Image> thumbnails, @Nullable final String uploader,
                          final String uploaderUrl, @NonNull final StreamType streamType) {
        this.title = name != null ? name : EMPTY_STRING;
        this.url = url != null ? url : EMPTY_STRING;
        this.serviceId = serviceId;
        this.duration = duration;
        this.thumbnails = thumbnails;
        this.uploader = uploader != null ? uploader : EMPTY_STRING;
        this.uploaderUrl = uploaderUrl;
        this.streamType = streamType;

        this.recoveryPosition = RECOVERY_UNSET;
    }

    @NonNull
    public String getTitle() {
        return title;
    }

    @NonNull
    public String getUrl() {
        return url;
    }

    public int getServiceId() {
        return serviceId;
    }

    public long getDuration() {
        return duration;
    }

    @NonNull
    public List<Image> getThumbnails() {
        return thumbnails;
    }

    @NonNull
    public String getUploader() {
        return uploader;
    }

    public String getUploaderUrl() {
        return uploaderUrl;
    }

    @NonNull
    public StreamType getStreamType() {
        return streamType;
    }

    public long getRecoveryPosition() {
        return recoveryPosition;
    }

    /*package-private*/ void setRecoveryPosition(final long recoveryPosition) {
        this.recoveryPosition = recoveryPosition;
    }

    @Nullable
    public Throwable getError() {
        return error;
    }

    @NonNull
    public Single<StreamInfo> getStream() {
        return ExtractorHelper.getStreamInfo(this.serviceId, this.url, false)
                .subscribeOn(Schedulers.io())
                .doOnError(throwable -> error = throwable);
    }

    public boolean isAutoQueued() {
        return isAutoQueued;
    }

    ////////////////////////////////////////////////////////////////////////////
    // Item States, keep external access out
    ////////////////////////////////////////////////////////////////////////////

    public void setAutoQueued(final boolean autoQueued) {
        isAutoQueued = autoQueued;
    }
}

File: app/src/main/java/org/schabi/newpipe/player/playqueue/PlayQueueItem.kt
package org.schabi.newpipe.player.playqueue

import java.io.Serializable
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.withContext
import org.schabi.newpipe.extractor.Image
import org.schabi.newpipe.extractor.stream.StreamInfo
import org.schabi.newpipe.extractor.stream.StreamInfoItem
import org.schabi.newpipe.extractor.stream.StreamType
import org.schabi.newpipe.util.ExtractorHelper

class PlayQueueItem : Serializable {
    val title: String
    val url: String
    val serviceId: Int
    val duration: Long
    val thumbnails: List<Image>
    val uploader: String
    val uploaderUrl: String?
    val streamType: StreamType

    var isAutoQueued: Boolean = false
    var recoveryPosition: Long = RECOVERY_UNSET
    var error: Throwable? = null
        private set

    constructor(info: StreamInfo) : this(
        info.name,
        info.url,
        info.serviceId,
        info.duration,
        info.thumbnails,
        info.uploaderName,
        info.uploaderUrl,
        info.streamType
    ) {
        if (info.startPosition > 0) {
            recoveryPosition = info.startPosition * 1000
        }
    }

    constructor(item: StreamInfoItem) : this(
        item.name,
        item.url,
        item.serviceId,
        item.duration,
        item.thumbnails,
        item.uploaderName,
        item.uploaderUrl,
        item.streamType
    )

    private constructor(
        name: String?,
        url: String?,
        serviceId: Int,
        duration: Long,
        thumbnails: List<Image>,
        uploader: String?,
        uploaderUrl: String?,
        streamType: StreamType
    ) {
        this.title = name ?: ""
        this.url = url ?: ""
        this.serviceId = serviceId
        this.duration = duration
        this.thumbnails = thumbnails
        this.uploader = uploader ?: ""
        this.uploaderUrl = uploaderUrl
        this.streamType = streamType
    }

    /**
     * Whether these two items should be treated as the same stream
     * for the sake of keeping the same player running when e.g. jumping between timestamps.
     *
     * @param other the [PlayQueueItem] to compare against.
     * @return whether the two items are the same so the stream can be re-used.
     */
    fun isSameItem(other: PlayQueueItem?): Boolean {
        if (other == null) return false
        return serviceId == other.serviceId && url == other.url
    }

    suspend fun getStream(): StreamInfo = withContext(Dispatchers.IO) {
        try {
            ExtractorHelper.getStreamInfo(serviceId, url, false)
        } catch (e: Throwable) {
            error = e
            throw e
        }
    }

    companion object {
        const val RECOVERY_UNSET = Long.MIN_VALUE
        private const val serialVersionUID = 1L
    }
}
